File: app/src/main/java/ai/elimu/content_provider/provider/SoundContentProvider.java
package ai.elimu.content_provider.provider;

import android.content.ContentProvider;
import android.content.ContentValues;
import android.content.Context;
import android.content.UriMatcher;
import android.database.Cursor;
import android.net.Uri;
import android.util.Log;

import java.util.List;

import ai.elimu.content_provider.BuildConfig;
import ai.elimu.content_provider.room.dao.SoundDao;
import ai.elimu.content_provider.room.db.RoomDb;

public class SoundContentProvider extends ContentProvider {

    private static final String AUTHORITY = BuildConfig.APPLICATION_ID + ".provider.sound_provider";
    private static final String TABLE_SOUNDS = "sounds";
    private static final int CODE_SOUNDS = 1;
    private static final int CODE_SOUND_ID = 2;

    private static final int CODE_SOUNDS_BY_LETTER_SOUND_ID = 3;
    private static final UriMatcher MATCHER = new UriMatcher(UriMatcher.NO_MATCH);

    static {
        MATCHER.addURI(AUTHORITY, TABLE_SOUNDS, CODE_SOUNDS);
        MATCHER.addURI(AUTHORITY, TABLE_SOUNDS + "/#", CODE_SOUND_ID);
        MATCHER.addURI(AUTHORITY, TABLE_SOUNDS + "/by-letter-sound-id/#", CODE_SOUNDS_BY_LETTER_SOUND_ID);
    }

    @Override
    public boolean onCreate() {
        Log.i(getClass().getName(), "onCreate");
        return true;
    }

    @Override
    public Cursor query(Uri uri, String[] projection, String selection, String[] selectionArgs, String sortOrder) {
        Log.i(getClass().getName(), "query");

        Log.i(getClass().getName(), "uri: " + uri);
        Log.i(getClass().getName(), "projection: " + projection);
        Log.i(getClass().getName(), "selection: " + selection);
        Log.i(getClass().getName(), "selectionArgs: " + selectionArgs);
        Log.i(getClass().getName(), "sortOrder: " + sortOrder);

        Context context = getContext();
        Log.i(getClass().getName(), "context: " + context);
        if (context == null) {
            return null;
        }

        RoomDb roomDb = RoomDb.getDatabase(context);
        SoundDao soundDao = roomDb.soundDao();

        final int code = MATCHER.match(uri);
        Log.i(getClass().getName(), "code: " + code);
        if (code == CODE_SOUNDS) {
            final Cursor cursor;

            // Get the Room Cursor
            cursor = soundDao.loadAllOrderedByUsageCount_Cursor();
            Log.i(getClass().getName(), "cursor: " + cursor);

            cursor.setNotificationUri(context.getContentResolver(), uri);

            return cursor;
        } else if (code == CODE_SOUNDS_BY_LETTER_SOUND_ID) {
            // Extract the letter-sound correspondence ID from the URI
            List<String> pathSegments = uri.getPathSegments();
            Log.i(getClass().getName(), "pathSegments: " + pathSegments);
            String letterSoundIdAsString = pathSegments.get(2);
            Long letterSoundId = Long.valueOf(letterSoundIdAsString);
            Log.i(getClass().getName(), "letterSoundId: " + letterSoundId);

            final Cursor cursor;

            // Get the Room Cursor
            cursor = soundDao.loadAllByLetterSound(letterSoundId);
            Log.i(getClass().getName(), "cursor: " + cursor);

            cursor.setNotificationUri(context.getContentResolver(), uri);

            return cursor;
        } else if (code == CODE_SOUND_ID) {
            // Extract the Sound ID from the URI
            List<String> pathSegments = uri.getPathSegments();
            Log.i(getClass().getName(), "pathSegments: " + pathSegments);
            String soundIdAsString = pathSegments.get(1);
            Long soundId = Long.valueOf(soundIdAsString);
            Log.i(getClass().getName(), "soundId: " + soundId);

            final Cursor cursor;

            // Get the Room Cursor
            cursor = soundDao.load_Cursor(soundId);
            Log.i(getClass().getName(), "cursor: " + cursor);

            cursor.setNotificationUri(context.getContentResolver(), uri);

            return cursor;
        } else {
            throw new IllegalArgumentException("Unknown URI: " + uri);
        }
    }

    @Override
    public String getType(Uri uri) {
        Log.i(getClass().getName(), "getType");

        throw new UnsupportedOperationException("Not yet implemented");
    }

    @Override
    public Uri insert(Uri uri, ContentValues values) {
        Log.i(getClass().getName(), "insert");

        throw new UnsupportedOperationException("Not yet implemented");
    }

    @Override
    public int update(Uri uri, ContentValues values, String selection, String[] selectionArgs) {
        Log.i(getClass().getName(), "update");

        throw new UnsupportedOperationException("Not yet implemented");
    }

    @Override
    public int delete(Uri uri, String selection, String[] selectionArgs) {
        Log.i(getClass().getName(), "delete");

        throw new UnsupportedOperationException("Not yet implemented");
    }
}

File: app/src/main/java/ai/elimu/content_provider/provider/SoundContentProvider.kt
package ai.elimu.content_provider.provider

import ai.elimu.content_provider.BuildConfig
import ai.elimu.content_provider.room.db.RoomDb
import android.content.ContentProvider
import android.content.ContentValues
import android.content.UriMatcher
import android.database.Cursor
import android.net.Uri
import android.util.Log

class SoundContentProvider : ContentProvider() {
    override fun onCreate(): Boolean {
        Log.i(javaClass.name, "onCreate")
        return true
    }

    override fun query(
        uri: Uri,
        projection: Array<String>?,
        selection: String?,
        selectionArgs: Array<String>?,
        sortOrder: String?
    ): Cursor? {
        Log.i(javaClass.name, "query")

        Log.i(javaClass.name, "uri: $uri")
        Log.i(javaClass.name, "projection: $projection")
        Log.i(javaClass.name, "selection: $selection")
        Log.i(javaClass.name, "selectionArgs: $selectionArgs")
        Log.i(javaClass.name, "sortOrder: $sortOrder")

        val context = context
        Log.i(javaClass.name, "context: $context")
        if (context == null) {
            return null
        }

        val roomDb = RoomDb.getDatabase(context)
        val soundDao = roomDb.soundDao()

        val code = MATCHER.match(uri)
        Log.i(javaClass.name, "code: $code")
        if (code == CODE_SOUNDS) {
            // Get the Room Cursor
            val cursor = soundDao.loadAllOrderedByUsageCount_Cursor()
            Log.i(javaClass.name, "cursor: $cursor")

            cursor.setNotificationUri(context.contentResolver, uri)

            return cursor
        } else if (code == CODE_SOUNDS_BY_LETTER_SOUND_ID) {
            // Extract the letter-sound correspondence ID from the URI
            val pathSegments = uri.pathSegments
            Log.i(javaClass.name, "pathSegments: $pathSegments")
            val letterSoundIdAsString = pathSegments[2]
            val letterSoundId = letterSoundIdAsString.toLong()
            Log.i(javaClass.name, "letterSoundId: $letterSoundId")

            // Get the Room Cursor
            val cursor = soundDao.loadAllByLetterSound(letterSoundId)
            Log.i(javaClass.name, "cursor: $cursor")

            cursor.setNotificationUri(context.contentResolver, uri)

            return cursor
        } else if (code == CODE_SOUND_ID) {
            // Extract the Sound ID from the URI
            val pathSegments = uri.pathSegments
            Log.i(javaClass.name, "pathSegments: $pathSegments")
            val soundIdAsString = pathSegments[1]
            val soundId = soundIdAsString.toLong()
            Log.i(javaClass.name, "soundId: $soundId")

            // Get the Room Cursor
            val cursor = soundDao.load_Cursor(soundId)
            Log.i(javaClass.name, "cursor: $cursor")

            cursor.setNotificationUri(context.contentResolver, uri)

            return cursor
        } else {
            throw IllegalArgumentException("Unknown URI: $uri")
        }
    }

    override fun getType(uri: Uri): String? {
        Log.i(javaClass.name, "getType")

        throw UnsupportedOperationException("Not yet implemented")
    }

    override fun insert(uri: Uri, values: ContentValues?): Uri? {
        Log.i(javaClass.name, "insert")

        throw UnsupportedOperationException("Not yet implemented")
    }

    override fun update(
        uri: Uri,
        values: ContentValues?,
        selection: String?,
        selectionArgs: Array<String>?
    ): Int {
        Log.i(javaClass.name, "update")

        throw UnsupportedOperationException("Not yet implemented")
    }

    override fun delete(uri: Uri, selection: String?, selectionArgs: Array<String>?): Int {
        Log.i(javaClass.name, "delete")

        throw UnsupportedOperationException("Not yet implemented")
    }

    companion object {
        private const val AUTHORITY = BuildConfig.APPLICATION_ID + ".provider.sound_provider"
        private const val TABLE_SOUNDS = "sounds"
        private const val CODE_SOUNDS = 1
        private const val CODE_SOUND_ID = 2

        private const val CODE_SOUNDS_BY_LETTER_SOUND_ID = 3
        private val MATCHER = UriMatcher(UriMatcher.NO_MATCH)

        init {
            MATCHER.addURI(AUTHORITY, TABLE_SOUNDS, CODE_SOUNDS)
            MATCHER.addURI(AUTHORITY, TABLE_SOUNDS + "/#", CODE_SOUND_ID)
            MATCHER.addURI(
                AUTHORITY,
                TABLE_SOUNDS + "/by-letter-sound-id/#",
                CODE_SOUNDS_BY_LETTER_SOUND_ID
            )
        }
    }
}
